### PR TITLE
Raise an explicit error message for Bundler 1.x

### PR DESCRIPTION
**Pull request by schneems**
With the Ruby 3.3.7 default, bundler 1.x no longer works on Heroku with the Ruby buildpack. Here's the details of what's causing the error:

The Ruby buildpack uses Ruby code to inspect the application at build time to know what dependencies and versions to install. The version of Ruby that the buildpack uses is tied to the default Ruby version. Recently this version was changed to Ruby 3.3.7 https://devcenter.heroku.com/changelog-items/3167.

The Ruby buildpack runs `bundle platform --ruby` in order to determine what Ruby version the application needs. This invokes bundler using the current Ruby version (3.3.7). However, bundler is using a method that was deprecated and removed in Ruby 3.2 https://ruby-doc.org/stdlib-2.7.1/libdoc/pathname/rdoc/Pathname.html#method-i-untaint. Because this method does not exist in Ruby 3.3 the bundler code cannot execute and fails with an error message:

```
remote:  !     There was an error parsing your Gemfile, we cannot continue
remote:  !     /tmp/d20250319-150-o0ictl/bundler-1.17.3/gems/bundler-1.17.3/lib/bundler/shared_helpers.rb:29:in `root': undefined method `untaint' for an instance of Pathname (NoMethodError)
remote:  !
remote:  !     Pathname.new(gemfile).untaint.expand_path.parent
remote:  !     ^^^^^^^^
remote:  !     from /tmp/d20250319-150-o0ictl/bundler-1.17.3/gems/bundler-1.17.3/lib/bundler.rb:234:in `root'
remote:  !     from /tmp/d20250319-150-o0ictl/bundler-1.17.3/gems/bundler-1.17.3/lib/bundler.rb:246:in `app_config_path'
remote:  !     from /tmp/d20250319-150-o0ictl/bundler-1.17.3/gems/bundler-1.17.3/lib/bundler.rb:273:in `settings'
remote:  !     from /tmp/d20250319-150-o0ictl/bundler-1.17.3/gems/bundler-1.17.3/lib/bundler/feature_flag.rb:21:in `block in settings_method'
remote:  !     from /tmp/d20250319-150-o0ictl/bundler-1.17.3/gems/bundler-1.17.3/lib/bundler/cli.rb:97:in `<class:CLI>'
remote:  !     from /tmp/d20250319-150-o0ictl/bundler-1.17.3/gems/bundler-1.17.3/lib/bundler/cli.rb:7:in `<module:Bundler>'
remote:  !     from /tmp/d20250319-150-o0ictl/bundler-1.17.3/gems/bundler-1.17.3/lib/bundler/cli.rb:6:in `<top (required)>'
remote:  !     from <internal:/tmp/tmp.CMaelujSAR/lib/ruby/3.3.0/rubygems/core_ext/kernel_require.rb>:136:in `require'
remote:  !     from <internal:/tmp/tmp.CMaelujSAR/lib/ruby/3.3.0/rubygems/core_ext/kernel_require.rb>:136:in `require'
remote:  !     from /tmp/tmp.CMaelujSAR/lib/ruby/gems/3.3.0/gems/bundler-2.5.22/exe/bundle:21:in `block in <top (required)>'
remote:  !     from /tmp/d20250319-150-o0ictl/bundler-1.17.3/gems/bundler-1.17.3/lib/bundler/friendly_errors.rb:124:in `with_friendly_errors'
remote:  !     from /tmp/tmp.CMaelujSAR/lib/ruby/gems/3.3.0/gems/bundler-2.5.22/exe/bundle:20:in `<top (required)>'
remote:  !     from /tmp/d20250319-150-o0ictl/bundler-1.17.3/bin/bundle:23:in `load'
remote:  !     from /tmp/d20250319-150-o0ictl/bundler-1.17.3/bin/bundle:23:in `<main>'
```

Bundler `1.17.3` was last released on December 27, 2018, and is no longer supported by bundler/rubygems core, so it has not been patched to remove this deprecated method. 

You can fix this issue by upgrading your application to bundler 2.3.x. I suggest `2.3.27`. Run:

```
$ gem install bundler -v 2.3.27
$ bundle update --bundler
```

Verify that the correct version is listed in the Gemfile.lock:

```
$ cat Gemfile.lock | grep "BUNDLED WITH" -A2
BUNDLED WITH
   2.3.27
```

And check the results into git:

```
$ git add Gemfile.lock
$ git commit -m "Update bundler version to 2.3.x"
```

Then deploy. Bundler 2.3.x supports Ruby >= 2.3.0 so it should work for your application. If you're using an older version of Rails that's locked to an older version of bundler, you can work around this with Rails LTS. That information and other upgrade suggestions are here https://www.reddit.com/r/Heroku/comments/1ij7b89/upgrading_ruby_versions_to_run_on_heroku24/.

## Additional concerns

Beyond `bundle platform --ruby` the buildpack also requires bundler internals to parse and load the Gemfile.lock. This strategy only works if the version of Ruby we're using for the buildpack is capable of running that version of bundler. Previously we only supported one version of bundler, but now that we support multiple, the Ruby version must be capable of running them all. We can move away from `bundle platform --ruby` by parsing the `Gemfile.lock` manually (this is what the CNB does) we can move away from requiring an arbitrary bundler version by listing out gems after they're installed via `gem list` or `bundle list`. However these are two relatively large behavior changes.

It's still possible to temporarily use bundler 1.17.3 on the platform by using an older buildpack version https://devcenter.heroku.com/articles/bundler-version#using-an-older-version-of-bundler. However this should only be considered a temporary fix and not a long term one.

### CODE REVIEW

File: spec/helpers/bundler_wrapper_spec.rb [hunk 1]
```diff
@@ -15,9 +15,9 @@
 describe "Bundler version detection" do
   it "supports minor versions" do
     wrapper_klass = LanguagePack::Helpers::BundlerWrapper
-    version = wrapper_klass.detect_bundler_version(contents: "BUNDLED WITH\n   1.17.3")
-    expect(wrapper_klass::BLESSED_BUNDLER_VERSIONS.key?("1")).to be_truthy
-    expect(version).to eq(wrapper_klass::BLESSED_BUNDLER_VERSIONS["1"])
+    expect {
+      wrapper_klass.detect_bundler_version(contents: "BUNDLED WITH\n   1.17.3")
+    }.to raise_error(BuildpackError)
```
[the remark this answers is not in the document]
Comment (schneems) [reply spec/helpers/bundler_wrapper_spec.rb:20]: My preference is to use "unit" tests for edge cases when possible. Because they're cheaper to run I don't feel bad about exercising all possible options versus hatchet tests are quite expensive. I deployed manually and verified that the exception is shown below. 

As a next step I'm looking into adding integration for `bin/report` to try and capture bundler versions in honeycomb. 

```
Buildpack set. Next release on glacial-sea-45203 will use https://github.com/heroku/heroku-buildpack-ruby#schneems/raise-nice-error-bundler-1.x.
$ cat Gemfile.lock
GEM
  remote: https://rubygems.org/
  specs:
    rake (13.2.1)

PLATFORMS
  arm64-darwin-24
  ruby

DEPENDENCIES
  rake

BUNDLED WITH
   1.17.2
$ git add .
$ git commit -m next
[main 16e73c1] next
 1 file changed, 1 insertion(+), 1 deletion(-)
$ time git push heroku
Enumerating objects: 7, done.
Counting objects: 100% (7/7), done.
Delta compression using up to 12 threads
Compressing objects: 100% (6/6), done.
Writing objects: 100% (7/7), 677 bytes | 677.00 KiB/s, done.
Total 7 (delta 1), reused 0 (delta 0), pack-reused 0 (from 0)
remote: Updated 2 paths from e94fca8
remote: Compressing source files... done.
remote: Building source:
remote:
remote: -----> Building on the Heroku-24 stack
remote: -----> Using buildpack: https://github.com/heroku/heroku-buildpack-ruby#schneems/raise-nice-error-bundler-1.x
remote: -----> Ruby app detected
remote:
remote:  !
remote:  !     Cannot install bundler 1.17.3
remote:  !
remote:  !     Your application requested bundler `1.17` in the `Gemfile.lock`
remote:  !     which resolved to `1.17.3`. This version no longer works on
remote:  !     Heroku. Please upgrade to bundler `2.3.x` or higher:
remote:  !
remote:  !     ```
remote:  !     $ gem install bundler -v 2.3.25
remote:  !     $ bundle update --bundler
remote:  !     $ git add Gemfile.lock
remote:  !     $ git commit -m "Updated bundler version"
remote:  !     ```
remote:  !
remote:  !     https://devcenter.heroku.com/changelog-items/3166
remote:  !
remote:  !     Push rejected, failed to compile Ruby app.
remote:
remote:  !     Push failed
remote:  !
remote:  ! ## Warning - The same version of this code has already been built: 16e73c1d161b69127ec967dfad027b9eabed3ec9
remote:  !
remote:  ! We have detected that you have triggered a build from source code with version 16e73c1d161b69127ec967dfad027b9eabed3ec9
remote:  ! at least twice. One common cause of this behavior is attempting to deploy code from a different branch.
remote:  !
remote:  ! If you are developing on a branch and deploying via git you must run:
remote:  !
remote:  !     git push heroku <branchname>:main
remote:  !
remote:  ! This article goes into details on the behavior:
remote:  !   https://devcenter.heroku.com/articles/duplicate-build-version
remote:
remote: Verifying deploy...
remote:
remote: !	Push rejected to glacial-sea-45203.
remote:
To https://git.heroku.com/glacial-sea-45203.git
 ! [remote rejected] main -> main (pre-receive hook declined)
error: failed to push some refs to 'https://git.heroku.com/glacial-sea-45203.git'
git push heroku  0.04s user 0.05s system 0% cpu 11.336 total
```

[the remark this answers is not in the document]
Comment (edmorley) [reply spec/helpers/bundler_wrapper_spec.rb:20]: > versus hatchet tests are quite expensive

A failing test is quite quick if the failure is right at the start of the build fwiw :-)
